Ensure a single handler manages channel's auto-read

lutovich opened the pull request:
`PullAllResponseHandler` is responsible for receiving a stream of records. Stream might be very large and handler tries to not fetch if fully into memory. It applies a network-level backpressure by disabling/enabling auto-read property of the underlying Netty channel. This property tells channel to not read from the network. Auto-read is disabled when number of queued records goes beyond the threshold.

Management of auto-read turned out to be problematic for nested queries within a single transaction. Nested queries result in multiple `PullAllResponseHandler`s being added to the queue of handlers. They will try to manage auto-read concurrently and can sometimes disable it completely. Callers would then be blocked and unable to proceed. This is not a problem for `Session#run()` because every such call ensures there was a logical SYNC and the previous query has completed and its result is buffered.

This PR fixes a problem by making only a single installed handler manage the channel's auto-read property. Making sure there only exists a single such handler is the responsibility of `InboundMessageDispatcher`. Auto-read is enabled when new auto-read managing handler is installed. Previous such handler is disabled.

Also renamed `InboundMessageDispatcher#queue()` to `InboundMessageDispatcher#enqueue()`.

Fixes https://github.com/neo4j/neo4j/issues/12033

lutovich committed:
Improve method name in InboundMessageDispatcher

lutovich committed:
Ensure a single handler manages channel's auto-read

`PullAllResponseHandler` is responsible for receiving a stream of
records. Stream might be very large and handler tries to not fetch
if fully into memory. It applies a network-level backpressure by
disabling/enabling auto-read property of the underlying Netty channel.
This property tells channel to not read from the network. Auto-read is
disabled when number of queued records goes beyond the threshold.

Management of auto-read turned out to be problematic for nested
queries within a single transaction. Nested queries result in
multiple `PullAllResponseHandler`s being added to the queue of
handlers. They will try to manage auto-read concurrently and can
sometimes disable it completely. Callers would then be blocked
and unable to proceed. This is not a problem for `Session#run()`
because every such call ensures there was a logical SYNC and the
previous query has completed and its result is buffered.

This commit fixes a problem by making only a single installed handler
manage the channel's auto-read property. Making sure there only exists
a single such handler is the responsibility of
`InboundMessageDispatcher`. Auto-read is enabled when new auto-read
managing handler is installed. Previous such handler is disabled.

ali-ince commented on driver/src/main/java/org/neo4j/driver/internal/async/inbound/InboundMessageDispatcher.java:
```diff
@@ -63,6 +66,7 @@ public void queue( ResponseHandler handler )
         else
         {
             handlers.add( handler );
+            updateAutoReadManagingHandlerIfNeeded( handler );
```
Even I couldn't think of a side effect, do you think that not cleaning `autoReadManagingHandler` after `handlers.remove()` would cause a problem?

lutovich replied:
Hm, I can't really see if there is a problem. However, it is probably incorrect to not clear `autoReadManagingHandler` field because `InboundMessageDispatcher` is pooled together with the channel. It's kind of strange to pool it together with some obsolete auto-read managing handler. Should be fixed in [this commit](https://github.com/neo4j/neo4j-java-driver/pull/528/commits/78cf1e3b5d27ee24862d0ec6811e41d83f3c95dd). Thanks!

ali-ince reviewed (commented):
looks good @lutovich. I only had one comment regarding the cleanup of field `autoReadManagingHandler`.

lutovich committed:
Forget auto-read managing handler when it is dequeued

Make sure `InboundMessageDispatcher` does not hold on to the last
auto-read management handler after it is removed from the queue of
all handlers.

lutovich committed:
Disable the removed auto-read managing handler

To make sure it does not influence the auto-read status of the channel.
Also restore default value of auto-read for channel when such handler
is removed.

This change is needed to avoid removing a handler that has just
disabled auto-read. This is possible when handler receives a RECORD
that causes it to disable auto-read and then receives a SUCCESS
message, which causes it to be dequeued with auto-read being disabled.
